src/transform/ReferenceCollection.js

```diff
@@ -0,0 +1,253 @@
+import NodeUtilities from './NodeUtilities'
+import Polyfill from './Polyfill'
+
+const REFERENCE_SELECTOR = '.reference'
+
+/**
+ * Is Citation.
+ * @param  {!string}  href
+ * @return {!boolean}
+ */
+const isCitation = href => href.indexOf('#cite_note') > -1
+
+/**
+ * Gets first child anchor.
+ * @param  {!HTMLElement} element
+ * @return {?HTMLAnchorElement}
+ */
+const getFirstChildAnchor = element => element.querySelector('A')
+
+/**
+ * Determines if node is a text node containing only whitespace.
+ * @param  {!Node}  node
+ * @return {!boolean}
+ */
+const isWhitespaceTextNode = node =>
+  !(!node || node.nodeType !== Node.TEXT_NODE || !node.textContent.match(/^\s+$/))
+
+/**
+ * Checks if element has a child anchor with a citation link.
+ * @param  {!HTMLElement}  element
+ * @return {!boolean}
+ */
+const hasCitationLink = element => {
+  try {
+    return isCitation(getFirstChildAnchor(element).getAttribute('href'))
+  } catch (e) {
+    return false
+  }
+}
+
+/**
+ * Get the reference text container.
+ * @param  {!Document} document
+ * @param  {!Node} sourceNode
+ * @return {?HTMLElement}
+ */
+const getRefTextContainer = (document, sourceNode) => {
+  const refTextContainerID = getFirstChildAnchor(sourceNode).getAttribute('href').slice(1)
+  const refTextContainer = document.getElementById(refTextContainerID)
+    || document.getElementById(decodeURIComponent(refTextContainerID))
+
+  return refTextContainer
+}
+
+/**
+ * Extract reference text free of backlinks.
+ * @param  {!Document} document
+ * @param  {!Node} sourceNode
+ * @return {!string}
+ */
+const collectRefText = (document, sourceNode) => {
+  const refTextContainer = getRefTextContainer(document, sourceNode)
+  if (!refTextContainer) {
+    return ''
+  }
+
+  // Clone what we're interested in into a frag so we can easily
+  // remove things without consequence to the 'live' document.
+  const frag = document.createDocumentFragment()
+  const fragDiv = document.createElement('div')
+  frag.appendChild(fragDiv)
+  // eslint-disable-next-line require-jsdoc
+  const cloneNodeIntoFragmentDiv = node => fragDiv.appendChild(node.cloneNode(true))
+  Array.prototype.slice.call(refTextContainer.childNodes)
+    .filter(NodeUtilities.isNodeTypeElementOrText)
+    .forEach(cloneNodeIntoFragmentDiv)
+
+  const removalSelector = 'sup[id^=cite_ref], .mw-cite-backlink'
+  Polyfill.querySelectorAll(fragDiv, removalSelector)
+    .forEach(node => node.remove())
+
+  return fragDiv.innerHTML.trim()
+}
+
+/**
+ * Get closest element to node which has class `reference`. If node itself has class `reference`
+ * returns the node.
+ * @param  {!Node} sourceNode
+ * @return {?HTMLElement}
+ */
+const closestReferenceClassElement = sourceNode => {
+  if (Polyfill.matchesSelector(sourceNode, REFERENCE_SELECTOR)) {
+    return sourceNode
+  }
+  return ElementUtilities.findClosestAncestor(sourceNode, REFERENCE_SELECTOR)
+}
+
+/**
+ * Reference item model.
+ */
+class ReferenceItem {
```

Comment:
What do you think about using a JSDoc for typing instead of a class? Then creating a ReferenceItem is just matching the shape which requires explicit labeling of properties. For example, before:

```
class ReferenceItem {
  constructor(id, rect, text, html) {
    this.id = id
    this.rect = rect
    this.text = text
    this.html = html
  }
}
const reference = new ReferenceItem('a', {width, height}, 'foo', '<b>bar</b>')
```

With labeled properties:
```
/**
 * @typedef ReferenceItem
 * @prop {string} id
 * @prop {DOMRect} rect
 * @prop {?string} text
 * @prop {?string} html
 */
const reference = {id: 'a', rect: {width, height}, text: 'foo', html: '<b>bar</b>'}
```

Same thing for NearbyReferences.

Reply:
> Usage of the class keyword exposes us to misusage of extends

Using classes to model data is pretty common. How is classes being extendable practically risky here?